// @azure/communication-email/dist-esm/src/generated/src/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { EmailMessage as EmailMessageMapper } from "../models/mappers";
export const endpoint = {
    parameterPath: "endpoint",
    mapper: {
        serializedName: "endpoint",
        required: true,
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export const messageId = {
    parameterPath: "messageId",
    mapper: {
        serializedName: "messageId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2021-10-01-preview",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String"
        }
    }
};
export const contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String"
        }
    }
};
export const emailMessage = {
    parameterPath: "emailMessage",
    mapper: EmailMessageMapper
};
export const repeatabilityRequestId = {
    parameterPath: "repeatabilityRequestId",
    mapper: {
        serializedName: "repeatability-request-id",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const repeatabilityFirstSent = {
    parameterPath: "repeatabilityFirstSent",
    mapper: {
        serializedName: "repeatability-first-sent",
        required: true,
        type: {
            name: "String"
        }
    }
};
